oscope.1
.\" @(#)$Id: oscope.1,v 2.2 2008/12/26 18:43:30 baccala Exp $
.\"
.\" Copyright (C) 1996 - 2000 Tim Witham <twitham@quiknet.com>
.\"
.\" (see the files README and COPYING for more details)

.TH OSCOPE 1 "May  6  2001" "Linux" "User Commands"
.SH NAME
oscope xoscope \- Digital Oscilloscope via Sound Card or Other Hardware
.SH SYNOPSIS
oscope [oscope options] [file]
.br
xoscope [toolkit options] [oscope options] [file]

.SH DESCRIPTION

.B Oscope
is a digital real-time oscilloscope. It graphically displays signal
amplitude or bit logic as a function of time.  Signals may be
displayed, saved, recalled, and manipulated by math functions.  Signal
input devices currently include:
.P

.TP 0.5i
.B /dev/dsp
Audio sound recording via /dev/dsp.  Two 8-bit analog channels at 8000
S/s to 44100 S/s.  Left and right audio is connected to X and Y inputs
respectively.  Use an external mixer program to select which sound
inputs to record.  AC coupled, voltages unknown, 256K sample memory.

.TP 0.5i
.B EsounD
Shared audio sound via the Enlightened Sound Daemon.  This is great
for watching music but support for it is an option at compile-time.
EsounD is auto-detected and preferred over /dev/dsp.

.TP 0.5i
.B ProbeScope / OsziFOX
Radio Shack ProbeScope, Cat. No. 22-310 is also known as an osziFOX.
This handheld probe sends its data through a serial port.  It samples
one channel at 6-bits up to 20 MS/s with 128 samples of memory.  Real
voltages are labeled in sample ranges from 1 volt to 100 volts.  If a
ProbeScope is detected, it is connected to the Z input.

.TP 0.5i
.B Bitscope
Bitscope (www.bitscope.com) is a mixed-signal capture engine which is
accessed through a serial port.  It simultaneously samples a digital
8-bit port and two analog channels at 8 bit resolution at up to 25
MS/s or more.  If detected, Channel A and B are connected to X and Y
while the Logic Analyzer is connected to Z, disabling all sound input.
Bitscope support is currently under development and not yet fully
functional.
.P

See the
.B -x
and
.B -z
options and the
.B ENVIRONMENT
section below for more details on how the above devices are detected.
Some of the controls below apply only to the sound card and are
labeled as such.
.B Oscope
has no physical control over the ProbeScope/osziFOX which is
controlled by its own switches and built-in menus.  Please refer to
your ProbeScope or osziFOX Owner's Manual for operating instructions.
Bitscope will eventually be controlled through a separate dialog
window.
.P

.PP
.SH "RUN\-TIME KEYBOARD CONTROLS"

.B Oscope
is an interactive program and can be completely controlled from the
keyboard at run-time.  In verbose key help mode, each available key is
shown on the screen in (parentheses).  The following single key
commands are available:

.TP 0.5i
.B ?
Toggle verbose key help display mode.

.TP 0.5i
.B Escape
Immediately quit the program.

.TP 0.5i
.B @
Load a previously saved file.  You are prompted for the filename.

.TP 0.5i
.B #
Save current settings and memory buffers to a file that can be loaded
later.  You are prompted for the filename and asked for confirmation
to overwrite if it already exists.

.TP 0.5i
.B Enter
Clear and refresh the entire screen.

.TP 0.5i
.B ^
Toggle the serial input device on/off.  A serial device is turned on
only if it is found active on a serial port.  This may be useful if
the device was not detected at startup.  See the
.B ENVIRONMENT
section below for more details.

.TP 0.5i
.B &
Cycle between the various input devices.  Note that this key will not
toggle to an unresponsive input device, so if only one device is
present, it will appear to have no effect.

.TP 0.5i
.B *
Different behavior for different input devices

For the soundcard, this key cycles the sound card DMA divisor: 4, 2,
1.  The sound driver will divide it's DMA buffer by this factor.  The
value 4 usually gives the fastest display rate.

Under EsounD, this value instead determines whether the connection to
EsounD will block (4) or not (2 or 1).  Blocking mode is nicest to CPU
usage but the
.B oscope
interface will not respond when the there is no sound stream coming
from EsounD.  Nonblocking mode will let
.B oscope
be responsive whether sound is available or not, but will consume all
available CPU cycles.

.TP 0.5i
.B (/)
Decrease/increase the sound card sampling rate.

.TP 0.5i
.B 9/0
Increase/decrease the Sec/Div horizontal time scale (zoom out/in on
time).

.TP 0.5i
.B -/=
Decrease/increase the sound card trigger level.

.TP 0.5i
.B _
Toggle the sound card trigger channel: X or Y.

.TP 0.5i
.B +
Cycle the sound card trigger type: automatic, rising edge, or falling
edge.

.TP 0.5i
.B Space
Cycle the trigger mode of the sound card: run, wait, stop.  Run mode
continuously acquires and displays samples after trigger events.  Wait
mode waits for the first trigger event and displays only the first set
of samples; this is "single-shot" mode.  Stop mode suspends the data
acquisition and displays the current samples.

.TP 0.5i
.B !
Cycle the plotting mode: point, point accumulate, line, or line
accumulate.  In the accumulate modes, all samples stay on the screen;
use
.B Enter
to clear them.

.TP 0.5i
.B ,
Cycle the graticule style: none, minor divisions only, or minor and
major divisions.

.TP 0.5i
.B .
Toggle the graticule position: behind or in front of the signals.

.TP 0.5i
.B '
Toggle the manual cursors on/off.  When manual cursors are displayed,
the measurements between the cursor positions are shown.  When cursors
are not displayed, automatic measurements are shown.

.TP 0.5i
.B """
Reset both manual cursor positions to the sample just after trigger.
.\" "

.TP 0.5i
.B Ctrl-q/w/e/r
The Control key held down in combination with q/w/e/r moves the first
cursor back or forward by 10 samples or back or forward by 1 sample
respectively.

.TP 0.5i
.B Ctrl-a/s/d/f
The Control key held down in combination with a/s/d/f moves the second
cursor back or forward by 10 samples or back or forward by 1 sample
respectively.

.TP 0.5i
.B 1\-8
Select the corresponding display channel.  Measurements are displayed
for the channel.  Channel 1 and 2 are used as input to the math
functions so they can't be used to do math.  By default, they are
connected to the X and Y input channels.  Channel 1 and 2 can also be
used to display memory buffers or the Z input for doing math on memory
or the alternate input.  Channel 3 through 8 are not restricted and
can be used for any purpose.  By default, the Z input is connected to
Channel 3.  The remaining single key commands operate on the currently
selected channel:

.TP 0.5i
.B Tab
Toggle visibility: Hide or show the selected channel.

.TP 0.5i
.B {/}
Decrease/Increase vertical scale of the selected channel.

.TP 0.5i
.B [/]
Decrease/Increase vertical position of the selected channel.

.TP 0.5i
.B `/~
Decrease/Increase number of logic analyzer bits displayed.  The
default of zero bits plots the signal as one analog line of varying
amplitude.  Any other value plots multiple digital lines representing
the least significant bits from bottom to top.

.TP 0.5i
.B ;/:
Increase/Decrease the math function of the selected channel.  This is
not available on channel 1 & 2.

.TP 0.5i
.B $
Show the result of an external math command on the selected channel.
You are prompted for the command.  The command must accept samples of
channel 1 & 2 on stdin and write a new signal to stdout.  See operl,
offt.c and xy.c in the distribution for examples of external math
filter commands.  Not available on channel 1 & 2.

.TP 0.5i
.B A-W
Store the currently selected channel into the corresponding memory
buffer.  Buffer X, Y and Z can not be used because they're reserved as
the signal inputs.  Memories are stored from time zero to the current
display update position.  So it is best to STOP the display before
storing to a memory buffer.

.TP 0.5i
.B a-z
Recall the corresponding memory buffer or input device to the
currently selected channel.  Buffer X is the Left or A input, Y is the
Right or B input, and Z is ProbeScope or Logic Analyzer input.  The
rest of the buffers are available for signal memory.

.PP
.SH "MOUSE CONTROLS"
.B Xoscope
adds mouse controls to menus or around the edges of the scope area.
These should be nearly self-explanatory.  They perform the same
functions as the equivalent keyboard commands above.  If built with
GTK+, a context-sensitive pop-up menu is available with right-click to
select channels, change scale and position, recall and store signals
and so on.  Left click decreases a variable while right click
increases.  The manual measurement cursors can also be positioned with
the mouse.

.PP
.SH "COMMAND\-LINE OPTIONS"

The command-line options define the startup state of
.B oscope
and have reasonable defaults.  All options may be capitalized in case
they conflict with an X toolkit option.  These options are also
recorded in text files saved by
.B oscope.

.TP 0.5i
.B -h
Help usage message showing these startup options with their default
values, then exit.

.TP 0.5i
.B -# <code>
Startup conditions of each channel.  # is a channel number from 1 to
8.  Code can have up to three fields, separated by colons:
position[.bits][:scale[:function #, memory letter, or external
command]].  Position is the number of pixels above (positive) or below
(negative) the center of the display.  Bits is the number of logic
analyzer bits to display.  Scale is a valid scaling factor from 1/50
to 50, expressed as a fraction.  The third field may contain a
built-in math function number, memory letter, or external math command
to run on the channel.  Using these options makes the channel visible
unless position begins with a '+', in which case the channel is
hidden.

.TP 0.5i
.B -a <channel>
Active, or selected, channel.

.TP 0.5i
.B -r <rate>
Sound card sampling Rate in samples per second.  Current valid values
are 8000, 11025, 22050, or 44100.

.TP 0.5i
.B -s <scale>
Time Scale factor from 1/20 to 1000 expressed as a fraction where 1/1
is 1 ms/div.

.TP 0.5i
.B -t <trigger>
Sound card Trigger conditions.  Trigger can have up to three fields,
separated by colons: position[:type[:channel]].  Position is the
number of pixels above (positive) or below (negative) the center of
the display.  Type is a number indicating the kind of trigger, 0 =
automatic, 1 = rising edge, 2 = falling edge.  Channel should be x or
y.

.TP 0.5i
.B -l <cursors>
Manual cursor Line positions.  Cursors can have up to three fields,
separated by colons: first[:second[:on?]].  First is the sample
position of the first cursor.  Second is the sample position of the
second cursor.  The final field is weather the manual cursors are
displayed (1) or the not displayed (0).

.TP 0.5i
.B -d <dma divisor>
Divisor for sound card DMA: 1, 2, or 4.  The sound driver will divide
it's DMA buffer by this factor.  The value 4 usually gives the fastest
display rate.

.TP 0.5i
.B -f
Font to use.  For
.B oscope,
these are listed in /usr/lib/kbd/consolefonts.  For
.B xoscope,
they're the output of xlsfonts.  The default should work best.

.TP 0.5i
.B -p <type>
Plot type.  0 = point, 1 = point accumulate, 2 = line, 3 = line
accumulate, 4 = step, 5 = step accumulate.

.TP 0.5i
.B -g <style>
Graticule style.  0 = none, 1 = minor divisions only, 2 = minor and
major divisions.

.TP 0.5i
.B -b
Whether the graticule is drawn Behind or in front of the signals.

.TP 0.5i
.B -v
Whether the Verbose key help is displayed.

.TP 0.5i
.B -x
Whether the sound card input device (XY) is turned on.  This can be
used to skip the attempt to connect to Esound or /dev/dsp.

.TP 0.5i
.B -z
Whether the serial input device (Z) is turned on.  This can be used to
suppress the search for a serial scope device.

.TP 0.5i
.B file
The name of a file to load upon startup.  This should be a file
previously saved by
.B oscope.

.SH EXAMPLES
.TP 0.5i
.B oscope -1 80 -2 -80 -3 0:1/5:6 -4 -160:1/5:7

This runs
.B oscope
with channel 1 above and channel 2 below the center of the display.
Also channel 3 and 4 are made visible to show the FFT of channel 1 and
2 respectively at a reduced scale of 1/5.

.TP 0.5i
.B xoscope oscope.dat

This runs xoscope, loading settings and memory buffers from a
previously saved data file called "oscope.dat".

.SH FILES
.B Oscope
creates readable text data files.  The files contain at least a
comment header which holds the current settings of
.B oscope.
Loading the file causes these saved settings to be restored.

To record your signals permanently first store them into memory
buffers, optionally recall them to channels, and then save the file.
All non-empty memory buffers are written to a column of the file
following the comment header.  Columns are separated by tab
characters.  These are stored back into the memory buffers when the
file is later loaded.  Simply recall them to channels to view them.
.P

This format could also be read by some spreadsheet or plotting
programs.  For example, the
.B gnuplot (1)
command
.P

plot "oscope.dat" using 0:1, "oscope.dat" using 0:2
.P

would plot the first and second columns of the "oscope.dat" data file.

.SH ENVIRONMENT

.TP 0.5i
.B OSCOPEPATH
The path to use when looking for external math commands.  If unset,
the built-in default is used.

.TP 0.5i
.B PROBESCOPE
The serial device your ProbeScope or osziFOX is connected to.  If
unset, /dev/probescope is used.  /dev/probescope could be a symbolic
link to the real device such as /dev/ttyS1.

.TP 0.5i
.B BITSCOPE
The serial device your Bitscope is connected to.  If unset,
/dev/bitscope is used.  /dev/bitscope could be a symbolic link to the
real device such as /dev/ttyS1.

.TP 0.5i
.B ESPEAKER
The host:port of the EsounD to connect to if built with EsounD
support.  If unset, localhost is assumed.  If no EsounD connection is
made or if there is no EsounD support compiled in, then
.B oscope
will try to read /dev/dsp directly.

.SH LIMITATIONS

The sound card should be capable of 44100 Hz sampling via the sound
drivers.  You must use an external mixer program to select the input
source device, level, etc.  Since these unknowns affect the amplitude,
there is no reference to voltage on the Y axis; it is in fact,
unknown.  Instead you're given the scale in pixels per sample unit.
Note that the serial oscilloscope devices don't have this limitation.
They have real voltage labels on the Y axis.
.P

Signal math is only valid if Channel 1 and 2 contain signals of the
same sampling rate.
.B It is up to you to make sure this is the case.  Doing math on signals
.B of different sample rates will produce incorrect results!
.P

The automatic measurements count zero crossings and divide to
determine the frequency and period.  If these zero crossings are not
"regularly-periodic", these measurements could be invalid.
.B Oscope
does understand how to measure the built-in FFT functions by locating
the peak frequency.  Use manual cursor positioning to get more precise
measurements.
.P

Your sound card is most-likely AC coupled so you will never see any DC
offset.  You probably can't get DC coupling by just shorting the input
capacitors on your sound card.  Use serial hardware to see DC offsets.
.P

The display may not be able to keep up if you give it too much to
plot, depending on your sound card, graphics card, and processor
speed.  External math commands are particularly expensive since the
kernel must then split the available CPU cycles across multiple
processes.  To maximize refresh speed, hide all unneeded channels, use
point or point accumulate mode, zoom in on Sec/Div as much as
possible, and turn off the graticule.
.P

Because it uses svgalib,
.B oscope
must be run as root or be setuid to root.
.B xoscope
doesn't have this restriction.

.SH BUGS
The keyboard interface may be confusing.

.SH AUTHOR
.B Oscope
was written by Tim Witham (twitham@quiknet.com), originally based
on "scope" by Jeff Tranter (Jeff_Tranter@Mitel.COM).
.B Oscope
is released under the conditions of the GNU General Public License.
See the files README and COPYING in the distribution for details.

.\" oscope.1 ends here.


oscope.c
/*
 * @(#)$Id: oscope.c,v 2.3 2008/12/28 04:31:33 baccala Exp $
 *
 * Copyright (C) 1996 - 2001 Tim Witham <twitham@quiknet.com>
 *
 * (see the files README and COPYING for more details)
 *
 * This file implements the basic oscilloscope internals
 *
 */

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <math.h>
#include "oscope.h"		/* program defaults */
#include "display.h"		/* display routines */
#include "func.h"		/* signal math functions */
#include "file.h"		/* file I/O functions */

/* global program structures */
Scope scope;

extern DataSrc datasrc_sc, datasrc_ps, datasrc_bs;
#ifdef HAVE_LIBESD
extern DataSrc datasrc_esd;
#endif
#ifdef HAVE_LIBCOMEDI
extern DataSrc datasrc_comedi;
#endif

DataSrc *datasrcs[] = {
#ifdef HAVE_LIBCOMEDI
  &datasrc_comedi,
#endif
#ifdef HAVE_LIBESD
  &datasrc_esd,
#endif
  &datasrc_sc,
  &datasrc_ps,
  &datasrc_bs
};

int ndatasrcs = sizeof(datasrcs)/sizeof(DataSrc *);
int datasrci = -1;
DataSrc *datasrc = NULL;

Channel ch[CHANNELS];

/* extra global variable definitions */
char *progname;			/* the program's name, autoset via argv[0] */
char version[] = VERSION;	/* version of the program, from Makefile */
char error[256];		/* buffer for "one-line" error messages */
int v_points;			/* pixels in vertical axis */
int h_points;			/* pixels in horizontal axis */
int offset;			/* vertical pixel offset to zero line */
int clip = 0;			/* whether we're maxed out or not */
char *filename;			/* default file name */
int frames = 0;			/* # of frames (full or partial) captured */
int in_progress = 0;		/* frame collection in progress?
				 *   if so, this is index of next sample
				 */

/* display command usage on stdout or stderr and exit */
void
usage(int error)
{
  static char *def[] = {"graticule", "signal"}, *onoff[] = {"on", "off"};
  FILE *where;

  where = error ? stderr : stdout;

  fprintf(where, "usage: %s [options] [file]\n\
\n\
Startup Options  Description (defaults)               version %s\n\
-h               this Help message and exit\n\
-D <datasrc>     select named data source (COMEDI/Soundcard/ESD/Bitscope/etc)\n\
-o <option>      specify data source specific options\n\
-# <code>        #=1-%d, code=pos[.bits][:scale[:func#, mem a-z or cmd]] (0:1/1)\n\
-a <channel>     set the Active channel: 1-%d                  (%d)\n\
-s <scale>       time Scale: 1/2000-1000/1 where 1=1ms/div    (%d/1)\n\
-t <trigger>     Trigger level[:type[:channel]]               (%s)\n\
-l <cursors>     cursor Line positions: first[:second[:on?]]  (%s)\n\
-f <font name>   the Font name as-in %s\n\
-p <type>        Plot mode: 0/1=point, 2/3=line, 4/5=step     (%d)\n\
-g <style>       Graticule: 0=none,  1=minor, 2=major         (%d)\n\
-i <min interv>  Minimum display update interval (ms)         (50)\n\
-b               %s Behind instead of in front of %s\n\
-v               turn Verbose key help display %s\n\
file             %s file to load to restore settings and memory\n\
",
	  progname, version, CHANNELS, CHANNELS, DEF_A,
	  DEF_S, DEF_T, DEF_L,
	  fonts,		/* the font method for the display */
	  scope.mode,
	  scope.grat, def[DEF_B], def[!DEF_B],
	  onoff[DEF_V], progname);
  exit(error);
}

/* handle command line options */
void
parse_args(int argc, char **argv)
{
  const char     *flags = "Hh"
    "1:2:3:4:5:6:7:8:"
    "a:r:s:t:l:c:m:d:f:p:g:o:i:bvxyz"
    "A:R:S:T:L:C:M:D:F:P:G:o:I:BVXYZ";
  int c;

  while ((c = getopt(argc, argv, flags)) != EOF) {
    handle_opt(c, optarg);
  }
}

/* cleanup before exiting due to error or program end */
void
cleanup()
{
  cleanup_math();
  cleanup_display();
}

/* initialize the scope */
void
init_scope()
{
  scope.mode = DEF_P;
  scope.scale = DEF_S;
  scope.div = 1;
  handle_opt('t', DEF_T);
  handle_opt('l', DEF_L);
  scope.grat = DEF_G;
  scope.behind = DEF_B;
  scope.run = 1;
  scope.select = DEF_A - 1;
  scope.verbose = DEF_V;
  scope.min_interval = 50;
}

/* initialize the signals */
void
init_channels()
{
  int i;
  static int first = 1;

  if (!first)
    do_math();			/* XXX halt currently running commands */
  first = 0;
  for (i = 0 ; i < CHANNELS ; i++) {
    bzero(&ch[i], sizeof(Channel));
    ch[i].signal = NULL;
    ch[i].target_mult = 1;
    ch[i].target_div = 1;
    ch[i].pos = 0;
    ch[i].show = 0;
    ch[i].bits = 0;
  }
}

/* samples needed to draw a current display of RATE */
int
samples(int rate)
{
  static unsigned long int r;
  r = (rate * ((h_points - 200) / 44) * scope.div / scope.scale / 1000) + 2;
  if (r > MAXWID) r = MAXWID;
  return (r);
}

/* scale num upward like a scope does, to max */
int
scaleup(int num, int max)
{
  int i;

  i = num;
  while (!(i % 10)) {
    i /= 10;
  }
  if (i == 2) num = num * 5 / 2;
  else num *= 2;
  if (num > max) num = max;
  return(num);
}

/* scale num downward like a scope does */
int
scaledown(int num)
{
  int i;

  i = num;
  while (!(i % 10)) {
    i /= 10;
  }
  if (i == 5) num = num * 2 / 5;
  else num /= 2;
  if (num < 1) num = 1;
  return(num);
}

/* Close the current data source */

void
datasrc_close(void)
{
  int j,k;

  if (datasrc) {

    if (datasrc->clear_trigger) datasrc->clear_trigger();
    scope.trige = 0;

    /* clear listeners on our signal channels prior to close */
    for (j=0; j<datasrc->nchans(); j++) {
      Signal *sig = datasrc->chan(j);
      if (sig->listeners) {
	for (k=0; k<CHANNELS; k++) {
	  if (ch[k].signal == sig) {
	    recall_on_channel(NULL, &ch[k]);
	  }
	}
      }
    }
  }

  datasrc = NULL;
  datasrci = -1;
}

int
datasrc_open(DataSrc *new_datasrc)
{
  int i;
  int limit = sizeof(datasrcs)/sizeof(DataSrc *);

  if (new_datasrc->nchans() > 0) {

    datasrc_close();

    datasrc = new_datasrc;

    for (i=0; i<limit; i++) {
      if (datasrc == datasrcs[i]) datasrci = i;
    }

    /* All data sources have at least one channel.  Show it. */

    if (ch[0].signal == NULL) {
      ch[0].show = 1;
      recall_on_channel(datasrc->chan(0), &ch[0]);
    }

    /* If data sources has a second channel, show it.  Older versions did. */

    if ((ch[1].signal == NULL) && (datasrc->nchans() > 1)) {
      ch[1].show = 1;
      recall_on_channel(datasrc->chan(1), &ch[1]);
    }

    return 1;

  }

  return 0;

}

/* Open a data source even if it's open function returns 0 */

void
datasrc_force_open(DataSrc *new_datasrc)
{
  int i;
  int limit = sizeof(datasrcs)/sizeof(DataSrc *);

  datasrc_close();

  datasrc = new_datasrc;

  for (i=0; i<limit; i++) {
    if (datasrc == datasrcs[i]) datasrci = i;
  }

  /* If data sources has a channel, show it. */

  /* XXX problem here - if data source requires options to
   * be set before it can open properly, nchans() won't
   * return anything valid yet
   */

  if ((ch[0].signal == NULL) && (datasrc->nchans() > 0)) {
    ch[0].show = 1;
    recall_on_channel(datasrc->chan(0), &ch[0]);
  }

  /* If data sources has a second channel, show it.  Older versions did. */

  if ((ch[1].signal == NULL) && (datasrc->nchans() > 1)) {
    ch[1].show = 1;
    recall_on_channel(datasrc->chan(1), &ch[1]);
  }

}

/* Find the first valid datasrc; should only be called once
 * return TRUE if successful; FALSE if no valid datasrcs were found
 */

int
datasrc_first(void)
{
  int i;
  int limit = sizeof(datasrcs)/sizeof(DataSrc *);

  for (i=0; i<limit; i++) {
    if (datasrc_open(datasrcs[i])) {
      return 1;
    }
  }
  return 0;
}

/* Advance to next datasrc
 * return TRUE if successful; FALSE if no other valid datasrcs were found
 */

int
datasrc_next(void)
{
  int i;
  int limit = sizeof(datasrcs)/sizeof(DataSrc *);

  if (datasrci < 0) return datasrc_first();

  for (i=(datasrci+1)%limit; (i=i%limit) != datasrci; i++) {
    if (datasrc_open(datasrcs[i])) {
      return 1;
    }
  }
  return 0;
}

/* Select the named datasrc (case insensitive match)
 * return TRUE if successful; FALSE if it wasn't found or couldn't be opened
 */

int
datasrc_byname(char *name)
{
  int i;
  int limit = sizeof(datasrcs)/sizeof(DataSrc *);

  /* Don't do anything if we're selecting the data source we've got */
  if (datasrc && strcasecmp(name, datasrc->name) == 0) return 1;

  for (i=0; i<limit; i++) {
    if (strcasecmp(name, datasrcs[i]->name) == 0) {
      datasrc_force_open(datasrcs[i]);
      return 1;
    }
  }

  return 0;
}

/* gr_* UIs call this after selecting file and confirming overwrite */
void
savefile(char *file)
{
  writefile(filename = file);
}

/* gr_* UIs call this after selecting file to load */
void
loadfile(char *file)
{
  readfile(filename = file);
}

/* handle single key commands */
void
handle_key(unsigned char c)
{
  static Channel *p;
  int displayed_samples, max_samples;

  p = &ch[scope.select];

  /* These next two are used for keyboard movement of the cursors.
   *
   * Cursors are stored as sample numbers (1 based).  max_samples
   * gives the largest legal value for a cursor.  displayed_samples
   * gives the number of samples displayed on the screen at once so we
   * can jump by 1/20 of a screen width.  If the signal is wider than
   * the screen width (i.e, the scrollbar is on), then these two
   * numbers will be different.
   */

  displayed_samples = p->signal ? samples(p->signal->rate) : 0;
  max_samples = p->signal ? max(samples(p->signal->rate), p->signal->num) : 0;

  if (c >= 'A' && c <= 'Z') {
    if (p->signal) {
      save(c);			/* store channel */
      clear();			/* need this in case other chan displays mem */
      draw_text(1);
    }
    return;
  } else if (c >= 'a' && c <= 'z') {
    if (datasrc && (c - 'a' < datasrc->nchans())) {
      recall(datasrc->chan(c - 'a'));	/* recall data channel */
    } else if (mem[c - 'a'].num > 0) {
      recall(&mem[c - 'a']);	/* recall memory location if something there */
    }
    p->show = 1;		/* always display newly recalled channel */
    clear();
    return;
  } else if (c >= '1' && c <= '0' + CHANNELS) {
    scope.select = (c - '1');	/* select channel */
    clear();			/* do this in case cursors move; see comments in display.c:drawdata() */
    return;
  }
  switch (c) {
  case 0:
  case -1:			/* no key pressed */
    break;
  case '\'':			/* toggle manual cursors */
    scope.curs = ! scope.curs;
    clear();
    break;
  case '"':			/* reset cursors to first sample */
    scope.cursa = scope.cursb = 1;
    clear();
    break;
  case 'q' - 96:		/* -96 is CTRL keys */
    if ((scope.cursa -= displayed_samples / 20) < 1)
      scope.cursa = max_samples - 1;
    break;
  case 'w' - 96:
    if ((scope.cursa += displayed_samples / 20) >= max_samples)
      scope.cursa = 1;
    break;
  case 'e' - 96:
    if (--scope.cursa < 1)
      scope.cursa = max_samples - 1;
    break;
  case 'r' - 96:
    if (++scope.cursa >= max_samples)
      scope.cursa = 1;
    break;
  case 'a' - 96:
    if ((scope.cursb -= displayed_samples / 20) < 1)
      scope.cursb = max_samples - 1;
    break;
  case 's' - 96:
    if ((scope.cursb += displayed_samples / 20) >= max_samples)
      scope.cursb = 1;
    break;
  case 'd' - 96:
    if (--scope.cursb < 1)
      scope.cursb = max_samples - 1;
    break;
  case 'f' - 96:
    if (++scope.cursb >= max_samples)
      scope.cursb = 1;
    break;
  case '\t':
    if (p->show) {		/* show / hide channel */
      p->show = 0;
    } else {
      p->show = 1;
    }
    clear();
    break;
  case '~':
    if ((p->bits += 2) > 16)
      p->bits = 0;
    clear();
    break;
  case '`':
    if ((p->bits -= 2) < 0)
      p->bits = 16;
    clear();
    break;
  case '}':
    if (p->target_div > 1)		/* increase scale */
      p->target_div = scaledown(p->target_div);
    else
      p->target_mult = scaleup(p->target_mult, 100);
    clear();
    break;
  case '{':
    if (p->target_mult > 1)		/* decrease scale */
      p->target_mult = scaledown(p->target_mult);
    else
      p->target_div = scaleup(p->target_div, 100);
    clear();
    break;
  case ']':
    p->pos -= 16;		/* position up */
    if (p->pos < -1 * v_points / 2)
      p->pos = v_points / 2;
    clear();
    break;
  case '[':
    p->pos += 16;		/* position down */
    if (p->pos > v_points / 2)
      p->pos = -1 * v_points / 2;
    clear();
    break;
  case ';':
    if (scope.select > 1) {	/* next math function */
      next_func();
      p->show = 1;
      clear();
    } else
      message("Math can not run on Channel 1 or 2", KEY_FG);
    break;
  case ':':
    if (scope.select > 1) {	/* previous math function */
      prev_func();
      p->show = 1;
      clear();
    } else
      message("Math can not run on Channel 1 or 2", KEY_FG);
    break;
  case '0':
    if (scope.div > 1)		/* decrease time scale, zoom in */
      scope.div = scaledown(scope.div);
    else
      scope.scale = scaleup(scope.scale, 5000);
    clear();
    break;
  case '9':
    if (scope.scale > 1)	/* increase time scale, zoom out */
      scope.scale = scaledown(scope.scale);
    else
      scope.div = scaleup(scope.div, 2000);
    clear();
    break;
  case '=':
    if (datasrc->set_trigger) {	/* increase trigger */
      scope.trig += 8;
      datasrc->set_trigger(scope.trigch, &scope.trig, scope.trige);
      clear();
    }
    break;
  case '-':
    if (datasrc->set_trigger) {	/* decrease trigger */
      scope.trig -= 8;
      datasrc->set_trigger(scope.trigch, &scope.trig, scope.trige);
      clear();
    }
    break;
  case '_':			/* change trigger channel */
    if (scope.trige != 0 && datasrc->set_trigger) {
      do {
	scope.trigch ++;
	if (scope.trigch >= datasrc->nchans()) scope.trigch = 0;
      } while (datasrc->set_trigger(scope.trigch,
				    &scope.trig, scope.trige) == 0);
      clear();
    }
    break;
  case '+':
    if (datasrc->set_trigger) {	/* change trigger type */
      do {
	scope.trige++;
	if (scope.trige > 2)
	  scope.trige = 0;
	if (scope.trige == 0 && datasrc->clear_trigger)
	  datasrc->clear_trigger();
      } while ((scope.trige != 0) &&
	       (datasrc->set_trigger(scope.trigch,
				     &scope.trig, scope.trige) == 0));
      clear();
    }
    break;
  case '(':
    if (datasrc && datasrc->change_rate(-1)) {
      in_progress = 0;
      clear();
    }
    break;
  case ')':
    if (datasrc && datasrc->change_rate(1)) {
      in_progress = 0;
      clear();
    }
    break;
  case '@':			/* load file */
    LoadSaveFile(0);
    break;
  case '#':			/* save file */
    LoadSaveFile(1);
    break;
  case '$':			/* run external math */
    if (scope.select > 1)
      ExternCommand();
    else
      message("External commands can not run on Channel 1 or 2", KEY_FG);
    break;
  case '&':			/* toggle between various data sources */
    if (datasrc_next()) {
      clear();
    }
    break;
  case '*':
    if (datasrc && (datasrc->option1 != NULL) && datasrc->option1()) {
      clear();
    }
    break;
  case '^':
    if (datasrc && (datasrc->option2 != NULL) && datasrc->option2()) {
      clear();
    }
    break;
  case '!':
    scope.mode++;		/* point, point accumulate, line, line acc. */
    if (scope.mode > 5)
      scope.mode = 0;
    clear();
    break;
  case ',':
    scope.grat++;		/* graticule off/on/more */
    if (scope.grat > 2)
      scope.grat = 0;
    clear();
    break;
  case '.':
    scope.behind = !scope.behind; /* graticule behind/in front of signal */
    draw_text(1);
    break;
  case '?':
  case '/':
    scope.verbose = !scope.verbose; /* on-screen help */
    clear();	/* XXX this is done to clear the help messages */
    break;
  case ' ':
    scope.run++;		/* run / wait / stop */
    if (scope.run > 2)
      scope.run = 0;
    if ((scope.run == 1) && datasrc)
      setinputfd(datasrc->fd());	/* were stopped, so now start */
    draw_text(1);
    break;
  case '\r':
  case '\n':
    clear();			/* refresh screen */
    break;
  case '\b':
  case 127:
    recall(NULL);		/* backspace/DEL - clear channel */
    clear();			/* otherwise chan freezes instead of clear */
    break;
  case '\e':
    cleanup();			/* quit */
    exit(0);
    break;
  default:
    c = 0;			/* ignore unknown keys */
  }
}

/* main program */
int
main(int argc, char **argv)
{
  progname = strrchr(argv[0], '/');
  if (progname == NULL)
    progname = argv[0];		/* global for error messages, usage */
  else
    progname++;
  init_scope();
  init_channels();
  init_math();
  if ((argc = OpenDisplay(argc, argv)) == FALSE)
    exit(1);
  parse_args(argc, argv);
  init_widgets();
  init_screen();
  /* clear(); */

  filename = FILENAME;
  if ((optind < argc) && (argv[optind] != NULL)) {
    filename = argv[optind];
    readfile(filename);
  } else if (!datasrc && ! datasrc_first()) {
    fprintf(stderr, "No valid data sources found... exiting\n");
#if 0
    exit(1);
#endif
  }
#if 0
  init_widgets();
  init_screen();
#endif
  clear();
  mainloop();
  cleanup();
  exit(0);
}

/* split_field() is used when we want to take a (possibly) long string
 * and split it across two fields.  'fieldtwo' is true to return field
 * 2, otherwise return field 1.
 */

char * split_field(char *str, int fieldtwo, int limit)
{
  static char buffer[256];
  char *sp_index = NULL;
  int i;

  if (strlen(str) > limit) {

    if (!fieldtwo) {

      for (i=0; i<limit; i++) {
	buffer[i] = str[i];

	if (str[i] == ' ') {
	  sp_index = &buffer[i];
	}
      }

      if (sp_index) {
	*sp_index = '\0';
      } else {
	buffer[i] = '\0';
      }

    } else {


      for (i=0; i<limit; i++) {

	if (str[i] == ' ') {
	  sp_index = &str[i];
	}
      }

      if (sp_index) {
	i = 0;
	while (*sp_index) buffer[i++] = *sp_index++;
	buffer[i] = '\0';
      } else {
	i = 0;
	while (str[limit+i]) buffer[i] = str[i];
	buffer[i] = '\0';
      }
    }

    return buffer;

  } else {

    /* Length of string less than limit */

    if (fieldtwo) return "";
    else return str;

  }
}
